### ISSUE TITLE
`lax.while_loop` accumulates numerical errata in 0.7.1, but not in 0.7.0, if PyTree is not 1D array

### ISSUE THREAD

**Issue by johannahaffner** [jax-ml/jax#31473]
### Description

Since 0.7.1, the while loop accumulates numerical errata. The final loss in the code below is orders of magnitude above the final loss in the previous version of JAX, 0.7.0. Disabling optimisation of the gradient computation restores the expected behaviour. 

This causes the Adam optimiser (e.g. in optax) to fail to find the unique global minimum of the Rosenbrock function within a very generous number of iterations, while it does find it with JAX 0.7.0.

I've left comments where simplifications to the code restore the expected behaviour. 

```python
import jax
import jax.numpy as jnp


jax.config.update("jax_enable_x64", True)  # Disabling double precision fixes it


def _tree_bias_correction(moment, decay, count):
    return jax.tree.map(lambda t: t / (1 - decay**count), moment)


def _tree_update_moment(updates, moments, decay):
    return jax.tree.map(
        lambda g, t: (1 - decay) * g + decay * t,
        updates,
        moments,
    )


def _tree_update_moment_per_elem_norm(updates, moments, decay):
    def orderth_norm(g):  # Moving this function definition into the lambda fixes it
        return g.conj() * g

    return jax.tree.map(
        lambda g, t: (1 - decay) * orderth_norm(g) + decay * t,
        updates,
        moments,
    )


def _init(params):
    mu, nu = jax.tree.map(lambda x: jnp.zeros_like(x), (params, params))
    return (jnp.zeros([], jnp.int32), mu, nu)


def _update(updates, state, params=None):
    del params
    count, mu, nu = state
    mu = _tree_update_moment(updates, mu, 0.9)
    nu = _tree_update_moment_per_elem_norm(updates, nu, 0.999)
    # Not calling bias correction fixes it
    mu_hat = _tree_bias_correction(mu, 0.9, count + 1)
    nu_hat = _tree_bias_correction(nu, 0.999, count + 1)
    updates = jax.tree.map(
        lambda m, v: m / (jnp.sqrt(v) + 1e-8),
        mu_hat,
        nu_hat,
    )
    return updates, (count + 1, mu, nu)


def rosenbrock(y):
    y_flat = y.flatten()
    residuals = (100 * (y_flat[1:] - y_flat[:-1]), 1.0 - y_flat[:-1])
    return 0.5 * jnp.sum(jnp.asarray(residuals) ** 2)


def _loop(init_y):
    learning_rate = 3e-3
    opt_state = _init(init_y)

    def cond_fun(carry):
        _, step, _, _ = carry
        return step < 10_000

    def body_fun(carry):
        y, step, opt_state, prev_loss = carry
        loss, grads = jax.value_and_grad(rosenbrock)(y)
        # grads = jax.lax.optimization_barrier(grads)  # Uncommenting fixes it
        updates, new_opt_state = _update(grads, opt_state, y)
        new_y = jax.tree.map(lambda _y, _u: _y - learning_rate * _u, y, updates)
        return (new_y, step + 1, new_opt_state, loss)

    init_state = (init_y, 0, opt_state, jnp.inf)
    final_state = jax.lax.while_loop(cond_fun, body_fun, init_state)
    final_y, final_steps, _, final_loss = final_state

    return final_y, final_steps, final_loss


init_y = 1.5 * jnp.ones((10, 2))
# init_y = 1.5 * jnp.ones((10, 1))  # Using a 1D array fixes it

final_y, steps, final_loss = _loop(init_y)
print(f"Final loss: {final_loss:.8f}, expected: ~0.0")
# 0.08863456 with JAX 0.7.1, 0.00000003 with 0.7.0
```

Cross-references: https://github.com/google-deepmind/optax/issues/1412, https://github.com/patrick-kidger/optimistix/issues/160

### System info (python version, jaxlib version, accelerator, etc.)

```
jax:    0.7.1
jaxlib: 0.7.1
numpy:  2.3.2
python: 3.13.3 (main, Apr  8 2025, 13:54:08) [Clang 16.0.0 (clang-1600.0.26.6)]
device info: cpu-1, 1 local devices"
process_count: 1
platform: uname_result(system='Darwin', node=REDACTED, release='24.5.0', version='Darwin Kernel Version 24.5.0: Tue Apr 22 19:54:26 PDT 2025; root:xnu-11417.121.6~2/RELEASE_ARM64_T8112', machine='arm64')
```

**Comment by WillFroom** [comment]
Hey @johannahaffner, thanks for the report, I bisected this issue down to https://github.com/openxla/xla/commit/7d7e3b2cbaca661a072d4b82fc3f6bb388d45866, which is a flag flip to the new XLA:CPU fusion emitter.

Fundamentally I _think_ that there isn't an issue with jax/XLA here but numerical instability in the example and you just got unlucky, which is why there is so many different ways to "fix" it: e.g. another change that "fixes" it on my machine is changing `lambda m, v: m / (jnp.sqrt(v) + 1e-8)` to `lambda m, v: m / (jnp.sqrt(v) + 1e-8 + 1e-15)`, i.e adding a tiny extra epsilon of `1e-15` which I guess just nudges the trajectory to avoid the numerical pathological case.

Hope this helps, let me know if anything isn't clear.

**Comment by johannahaffner** [comment]
Thanks for taking a look @WillFroom!

What was puzzling to me is that I would expect the solver to be better at avoiding numerical pathologies in higher precision, not worse, but while that may be true in general we might indeed capture one specific pathological point in double precision and then Adam will carry this around for a while. That and the shape dependency prompted me to investigate this more deeply. 

Ultimately, I think you're right - the fact that so many different "fixes" exist could be explained by a common denominator that is specific to Rosenbrock, with this initialisation + solver.
